From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390: report spectre mitigation via syslog
Patch-mainline: not yet, under development
References: bnc#1089386, LTC#166572

Description:  kernel: expoline defense against the spectre attack
Symptom:      None
Problem:      The spectre attack may be used to read restricted kernel
              data from user space.
Solution:     Provide execute trampolines for indirect branches
Reproduction: None

Upstream-Description:

              s390: report spectre mitigation via syslog

              Add a boot message if either of the spectre defenses is active.
              The message is
                  "Spectre V2 mitigation: execute trampolines."
              or  "Spectre V2 mitigation: limited branch prediction."

              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/kernel/nospec-branch.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/arch/s390/kernel/nospec-branch.c
+++ b/arch/s390/kernel/nospec-branch.c
@@ -41,6 +41,18 @@ static int __init nogmb_setup_early(char
 }
 early_param("nogmb", nogmb_setup_early);
 
+static int __init nospec_report(void)
+{
+#ifdef CC_USING_EXPOLINE
+	if (!nospec_disable)
+		pr_info("Spectre V2 mitigation: execute trampolines.\n");
+#endif
+	if (__test_facility(82, S390_lowcore.alt_stfle_fac_list))
+		pr_info("Spectre V2 mitigation: limited branch prediction.\n");
+	return 0;
+}
+arch_initcall(nospec_report);
+
 #ifdef CONFIG_EXPOLINE
 
 #ifdef CONFIG_EXPOLINE_OFF
